core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,197 @@
+  'ExplorationPlayerStateService', 'UrlInterpolationService',
+  'ComputeGraphService', 'ExtractImageFilenamesFromStateService',
+  function($q, $uibModal, ExplorationContextService, AssetsBackendApiService,
+      ExplorationPlayerStateService, UrlInterpolationService,
+      ComputeGraphService, ExtractImageFilenamesFromStateService) {
+    var MAX_NUM_IMAGE_FILES_TO_DOWNLOAD_SIMULTANEOUSLY = 3;
+
+    var _filenamesOfImageCurrentlyDownloading = [];
+    var _filenamesOfImageToBeDownloaded = [];
+    var _exploration = null;
+    // imageLoadedCallback is an object of objects (identified by the filenames
+    // which are being downloaded at the time they are required by the
+    // directive).The object contains the resolve method of the promise
+    // attached with getInImageUrl method.
+    var _imageLoadedCallback = {};
+
+    var _init = function(exploration) {
+      _exploration = exploration;
+      _states = exploration.states;
+    };
+
+    /**
+     * Gets the Url for the image file.
+     * @param {string} filename - Filename of the image whose Url is to be
+     *                            created.
+     * @param {function} onLoadCallback - Function that is called when the
+     *                                    Url of the loaded image is obtained.
+     */
+    var _getUrlUsingFileInCache = function(filename, onLoadCallback) {
+      AssetsBackendApiService.loadImage(
+        ExplorationContextService.getExplorationId(), filename)
+        .then(function(loadedImageFile) {
+          var objectUrl = URL.createObjectURL(loadedImageFile.data);
+          if (onLoadCallback) {
+            onLoadCallback(objectUrl);
+          }
+        });
+    };
+
+    /**
+     * Gets image files in Bfs order from the state.
+     * @param {string} sourceStateName - The name of the state starting from
+     *                                   which the filenames should be
+     *                                   obtained.
+     */
+    var _getImageFilenamesInBfsOrder = function(sourceStateName) {
+      var stateNamesInBfsOrder = (
+        ComputeGraphService.computeBfsTraversalOfStates(
+          _exploration.getInitialState().name,
+          _exploration.getStates(),
+          sourceStateName));
+      var imageFilenames = [];
+
+      stateNamesInBfsOrder.forEach(function(stateName) {
+        var state = _states.getState(stateName);
+        ExtractImageFilenamesFromStateService.getImageFilenamesInState(state).
+          forEach(function(filename) {
+            imageFilenames.push(filename);
+          });
+      });
+      return imageFilenames;
+    };
+
+    /**
+     * Handles the loading of the image file.
+     * @param {string} imageFilename - The filename of the image to be loaded.
+     */
+    var _loadImage = function(imageFilename) {
+      AssetsBackendApiService.loadImage(
+        ExplorationContextService.getExplorationId(), imageFilename
```

Comment:
Can we get the exploration id from the _exploration object?

Reply:
Ditto as above.